lib/mempool/indexer.js

```diff
@@ -0,0 +1,483 @@
+   * @param {MempoolEntry} entry
+   */
+
+  removeEntry(entry) {
+    const tx = entry.tx;
+    const hash = tx.hash();
+
+    this.txIndex.remove(hash);
+
+    for (const {prevout} of tx.inputs) {
+      const {hash, index} = prevout;
+      const prev = this.mempool.getTX(hash);
+
+      if (!prev)
+        continue;
+
+      this.coinIndex.insert(prev, index);
+    }
+
+    for (let i = 0; i < tx.outputs.length; i++)
+      this.coinIndex.remove(hash, i);
+  }
+
+  /**
+   * Transaction was included in the block.
+   * We want to get rid of the input coins as well as output coins.
+   * (if there are any)
+   * This event comes after remove entry, which recovers inputs
+   * as coins, we want to get rid of those coins as well.
+   * @param {TX} tx
+   * @param {Block} block
+   */
+
+  confirmed(tx, block) {
+  }
+
+  /**
+   * Block disconnected and we recover coins if they are available.
+   * This event comes after `tx` and `add entry`, we might want to
+   * check if outputs indexed by those are already spent in the mempool.
+   * @param {TX} tx
+   * @param {Block} block
+   */
+
+  unconfirmed(tx, block) {
+  }
+
+  /**
+   * Transaction was double spent in the mempool.
+   * We want to recover coisn that are not spent in the mempool.
+   * @param {MempoolEntry} entry
+   */
+
+  doubleSpend(entry) {
+  }
+
+  /**
+   * Reset indexes
+   * @private
+   */
+
+  reset() {
+    this.txIndex.reset();
+    this.coinIndex.reset();
+  }
+
+  /**
+   * Find all transactions partaining to a certain address.
+   * Note: this does not accept multiple addresses.
+   * @param {Address} addr
+   * @returns {TX[]}
+   */
+
+  getTXByAddress(addr) {
+    const hash = Address.getHash(addr);
+
+    return this.txIndex.get(hash);
+  }
+
+  /**
+   * Find all transactions pertaining to a certain address.
+   * @param {Address} addr
+   * @param {TXMeta[]]}
+   */
+
+  getMetaByAddress(addr) {
+    const hash = Address.getHash(addr);
+
+    return this.txIndex.getMeta(hash);
+  }
+
+  /**
+   * Find all coins pertaining to a certain address.
+   * @param {Address} addr
+   * @return {Coin[]}
+   */
+
+  getCoinsByAddress(addr) {
+    const hash = Address.getHash(addr);
+
+    return this.coinIndex.get(hash);
+  }
+}
+
+/**
+ * Mempool Indexer Options
+ * @alias module:mempool.MempoolIndexerOptions
+ */
+
+class MempoolIndexerOptions {
+  /**
+   * Create indexer options.
+   * @param {Object}
+   */
+
+  constructor(options) {
+    this.mempool = null;
+
+    this.fromOptions(options);
+  }
+
+  /**
+   * Inject properties from object.
+   * @private
+   * @param {Object} options
+   * returns {MempoolIndexerOptions}
+   */
+
+  fromOptions(options) {
+    assert(options, 'Mempool indexer requires options.');
+    assert(options.mempool && typeof options.mempool === 'object',
+      'Mempool indexer requires a mempool.'
+    );
+
+    this.mempool = options.mempool;
+
+    return this;
+  }
+}
+
+/**
+ * TX Address Index
+ * @ignore
+ */
+
+class TXIndex {
```

Comment:
This is a TxIndex specific to the mempool right? No relation to the main TXIndexer?

Reply:
Nope, at least for now it's not related to the TXIndexer in a direct way, even though this is doing same thing just in mempool.